File: src/improvement/test_quality.py
"""Test quality analysis — beyond coverage.

Detects:
- Weak assertions (assert True, assert x is not None without value check)
- Tests with no assertions
- Duplicate test names
"""

from __future__ import annotations

import ast
import os
from dataclasses import dataclass


@dataclass
class TestQualityFinding:
    file: str
    test_name: str
    line: int
    category: str  # "weak_assertion", "no_assertion", "duplicate_name"
    description: str


WEAK_ASSERTIONS = {
    "assertTrue": "Consider asserting a specific value instead of True",
    "assertIsNotNone": "Consider asserting the specific expected value",
    "assertFalse": "Consider asserting a specific condition",
}


def find_weak_assertions(filepath: str) -> list[TestQualityFinding]:
    """Find tests with weak assertions."""
    try:
        with open(filepath) as f:
            source = f.read()
        tree = ast.parse(source)
    except (SyntaxError, FileNotFoundError, OSError):
        return []

    findings = []
    for node in ast.walk(tree):
        if not isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
            continue
        if not node.name.startswith("test_"):
            continue

        for child in ast.walk(node):
            if isinstance(child, ast.Call) and isinstance(child.func, ast.Attribute):
                method = child.func.attr
                if method in WEAK_ASSERTIONS:
                    findings.append(TestQualityFinding(
                        file=filepath,
                        test_name=node.name,
                        line=child.lineno,
                        category="weak_assertion",
                        description=f"{method}() in {node.name}: {WEAK_ASSERTIONS[method]}",
                    ))

    return findings


def find_tests_without_assertions(filepath: str) -> list[TestQualityFinding]:
    """Find test functions that contain no assert statements."""
    try:
        with open(filepath) as f:
            source = f.read()
        tree = ast.parse(source)
    except (SyntaxError, FileNotFoundError, OSError):
        return []

    findings = []
    for node in ast.walk(tree):
        if not isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
            continue
        if not node.name.startswith("test_"):
            continue

        has_assert = False
        for child in ast.walk(node):
            if isinstance(child, ast.Assert):
                has_assert = True
                break
            if isinstance(child, ast.Call) and isinstance(child.func, ast.Attribute):
                if child.func.attr.startswith("assert"):
                    has_assert = True
                    break
            # pytest.raises counts as an assertion
            if isinstance(child, ast.Call) and isinstance(child.func, ast.Attribute):
                if child.func.attr == "raises":
                    has_assert = True
                    break

        if not has_assert:
            findings.append(TestQualityFinding(
                file=filepath,
                test_name=node.name,
                line=node.lineno,
                category="no_assertion",
                description=f"Test '{node.name}' has no assertions",
            ))

    return findings


def find_duplicate_test_names(test_dir: str) -> list[TestQualityFinding]:
    """Find duplicate test function names across the test suite."""
    seen: dict[str, str] = {}  # name → first file
    findings = []

    for root, dirs, files in os.walk(test_dir):
        dirs[:] = [d for d in dirs if d != "__pycache__"]
        for name in sorted(files):
            if not name.startswith("test_") or not name.endswith(".py"):
                continue
            filepath = os.path.join(root, name)
            try:
                with open(filepath) as f:
                    tree = ast.parse(f.read())
            except (SyntaxError, OSError):
                continue

            for node in ast.walk(tree):
                if not isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
                    continue
                if not node.name.startswith("test_"):
                    continue
                if node.name in seen and seen[node.name] != filepath:
                    findings.append(TestQualityFinding(
                        file=filepath,
                        test_name=node.name,
                        line=node.lineno,
                        category="duplicate_name",
                        description=f"Test '{node.name}' also exists in {seen[node.name]}",
                    ))
                else:
                    seen[node.name] = filepath

    return findings


def scan_test_quality(test_dir: str) -> list[TestQualityFinding]:
    """Run all test quality checks."""
    findings = []

    for root, dirs, files in os.walk(test_dir):
        dirs[:] = [d for d in dirs if d != "__pycache__"]
        for name in sorted(files):
            if not name.startswith("test_") or not name.endswith(".py"):
                continue
            filepath = os.path.join(root, name)
            findings.extend(find_weak_assertions(filepath))
            findings.extend(find_tests_without_assertions(filepath))

    findings.extend(find_duplicate_test_names(test_dir))
    return findings

File: rust/src/improvement/test_quality.rs
//! Test quality analysis — weak assertions, no-assertion tests, duplicate names.

use regex::Regex;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// A test quality finding.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TestQualityFinding {
    pub file: String,
    pub test_name: String,
    pub line: usize,
    pub category: String,
    pub description: String,
}

/// Known weak assertion methods and their suggestions.
pub const WEAK_ASSERTIONS: &[(&str, &str)] = &[
    (
        "assertTrue",
        "Consider asserting a specific value instead of True",
    ),
    (
        "assertIsNotNone",
        "Consider asserting the specific expected value",
    ),
    (
        "assertFalse",
        "Consider asserting a specific condition",
    ),
];

/// Find weak assertions in test source code.
pub fn find_weak_assertions(filepath: &str, source: &str) -> Vec<TestQualityFinding> {
    let mut findings = Vec::new();
    let test_fn_re = Regex::new(r"^\s*(async\s+)?def\s+(test_\w+)").unwrap();

    let lines: Vec<&str> = source.lines().collect();
    let mut current_test: Option<(String, usize)> = None;

    for (i, line) in lines.iter().enumerate() {
        let line_num = i + 1;

        if let Some(caps) = test_fn_re.captures(line) {
            current_test = Some((caps[2].to_string(), line_num));
        }

        if let Some((ref test_name, _)) = current_test {
            for &(method, suggestion) in WEAK_ASSERTIONS {
                if line.contains(method) {
                    findings.push(TestQualityFinding {
                        file: filepath.to_string(),
                        test_name: test_name.clone(),
                        line: line_num,
                        category: "weak_assertion".into(),
                        description: format!(
                            "{}() in {}: {}",
                            method, test_name, suggestion
                        ),
                    });
                }
            }
        }
    }
    findings
}

/// Find test functions that contain no assert statements.
pub fn find_tests_without_assertions(filepath: &str, source: &str) -> Vec<TestQualityFinding> {
    let mut findings = Vec::new();
    let test_fn_re = Regex::new(r"^\s*(async\s+)?def\s+(test_\w+)").unwrap();

    let lines: Vec<&str> = source.lines().collect();
    let mut i = 0;
    while i < lines.len() {
        if let Some(caps) = test_fn_re.captures(lines[i]) {
            let test_name = caps[2].to_string();
            let start_line = i + 1;
            let indent = lines[i].len() - lines[i].trim_start().len();

            // Collect body
            let mut end = i + 1;
            while end < lines.len() {
                let line = lines[end];
                if !line.trim().is_empty() {
                    let cur_indent = line.len() - line.trim_start().len();
                    if cur_indent <= indent && test_fn_re.is_match(line) {
                        break;
                    }
                }
                end += 1;
            }

            let body = lines[i..end].join("\n");
            let has_assert = body.contains("assert")
                || body.contains(".raises(")
                || body.contains("assert_eq!")
                || body.contains("assert_ne!");

            if !has_assert {
                findings.push(TestQualityFinding {
                    file: filepath.to_string(),
                    test_name: test_name.clone(),
                    line: start_line,
                    category: "no_assertion".into(),
                    description: format!("Test '{}' has no assertions", test_name),
                });
            }
            i = end;
        } else {
            i += 1;
        }
    }
    findings
}

/// Find duplicate test function names across multiple files.
pub fn find_duplicate_test_names(
    file_tests: &[(String, Vec<(String, usize)>)],
) -> Vec<TestQualityFinding> {
    let mut seen: HashMap<String, String> = HashMap::new();
    let mut findings = Vec::new();

    for (filepath, tests) in file_tests {
        for (test_name, line) in tests {
            if let Some(first_file) = seen.get(test_name) {
                if first_file != filepath {
                    findings.push(TestQualityFinding {
                        file: filepath.clone(),
                        test_name: test_name.clone(),
                        line: *line,
                        category: "duplicate_name".into(),
                        description: format!(
                            "Test '{}' also exists in {}",
                            test_name, first_file
                        ),
                    });
                }
            } else {
                seen.insert(test_name.clone(), filepath.clone());
            }
        }
    }
    findings
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_find_weak_assertions() {
        let src = "def test_foo():\n    self.assertTrue(result)\n";
        let findings = find_weak_assertions("test_x.py", src);
        assert_eq!(findings.len(), 1);
        assert_eq!(findings[0].category, "weak_assertion");
    }

    #[test]
    fn test_find_tests_without_assertions() {
        let src = "def test_noop():\n    x = 1\n    y = 2\n";
        let findings = find_tests_without_assertions("test_x.py", src);
        assert_eq!(findings.len(), 1);
        assert_eq!(findings[0].category, "no_assertion");
    }

    #[test]
    fn test_find_tests_with_assertions() {
        let src = "def test_good():\n    assert x == 1\n";
        let findings = find_tests_without_assertions("test_x.py", src);
        assert!(findings.is_empty());
    }

    #[test]
    fn test_find_duplicate_test_names() {
        let file_tests = vec![
            (
                "test_a.py".to_string(),
                vec![("test_foo".to_string(), 1)],
            ),
            (
                "test_b.py".to_string(),
                vec![("test_foo".to_string(), 5)],
            ),
        ];
        let findings = find_duplicate_test_names(&file_tests);
        assert_eq!(findings.len(), 1);
        assert_eq!(findings[0].category, "duplicate_name");
    }
}
